### PR TITLE
Respect cacheResult flag in safe initialization cache.

### PR DESCRIPTION
**Pull request by q-ata**
The safe initialization checker only needs to cache entries on certain `cachedEval` calls, specified by a `cacheResult` flag. This PR reduces the size of the safe initialization cache during compilation with `-Ysafe-init` by respecting the value of the flag, and only caching when it is true.

### CODE REVIEW

**Comment by q-ata** [comment]
RE: Signed the Scala CLA.

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 1]
```diff
@@ -893,7 +893,7 @@ object Semantic:
 
               case Cold => Cold
 
-              case ref: Ref => eval(vdef.rhs, ref, enclosingClass)
+              case ref: Ref => eval(vdef.rhs, ref, enclosingClass, cacheResult = true)
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:896]: Please add test cases that motivated the two changes in `Semantic.scala`.

**Review by olhotak — COMMENTED** [review]
Otherwise LGTM.

**Review by olhotak — APPROVED** [review]

File: compiler/src/dotty/tools/dotc/transform/init/Semantic.scala [hunk 2]
```diff
@@ -989,7 +989,7 @@ object Semantic:
             val errors = Reporter.stopEarly {
               val res = {
                 given Trace = Trace.empty
-                eval(body, thisV, klass)
+                eval(body, thisV, klass, cacheResult = true)
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/init/Semantic.scala:992]: BTW, it seems we miss a test for this one.

**Pull request merged**